cni/pkg/plugin/plugin.go

```diff
@@ -220,7 +220,7 @@ func CmdAdd(args *skel.CmdArgs) (err error) {
 
 			log.Debugf("ambientConf.ZTunnelReady: %v", ambientConf.ZTunnelReady)
 			added := false
-			if !excludePod && ambientConf.ZTunnelReady {
```

Comment:
If ztunnel is crashing, this change will drop the packet directly : https://github.com/istio/istio/pull/44208/files#diff-c3b29496fc3e6f9e25813ee08e098497adfa7061d3d3e24582892c39996183f8R135  What do you mean ' keep forwarding`?

Reply:
What I mean by "latching" is once we program redirection we should not turn it off unless the user took an explicit action to uninstall.

0. Nothing installed
1. Ztunnel, CNI deployed but not yet `Ready` - we should have no rules
2. Ztunnel becomes ready - we should program redirection
3. Ztunnel becomes unready - we should NOT undo redirection

Reply:
I guess I'm still not convinced we really need to care about zt readiness as a redirection rule gate tho. It's simpler to just let CNI do what it needs to do as soon as it knows it will need to.

That way _as soon as_ the user indicates to the control plane that redirection is desired we start redirecting. If ztunnel isn't ready to proxy then the traffic will be dropped, but that seems less surprising than not redirecting at all until ztunnel is ready, whenever that happens to be.

The main benefit of tying CNI readiness and ztunnel readiness together into a single state is it allows for e.g. node scheduling taints as mentioned here: https://github.com/istio/istio/issues/40303 to work more simply, I guess.

Reply:
My concern with that is if you just do `istioctl install` then the redirect starts happening before ztunnel is deployed and the install is not zero-downtime.

Reply:
That at least is something I would be OK fixing with an out of band "wait for ztunnels" `istioctl install` gate, like we do already for ingressgateways/CNI/etc.

Reply:
I think helm has a --wait and we can specify to install ztunnel first probably.

I am not sure this solves "new nodes" issues. We might have other issues for new nodes though, since we lose the ability to use our workaround for the race in https://github.com/istio/istio/issues/40303

Reply:
I'm not very sure how to understand 

> we lose the ability to use our workaround for the race in https://github.com/istio/istio/issues/40303

I think https://github.com/istio/istio/issues/40303 is to guarantee `istio-cni` is ready before pods being scheduled on that node?

1. New node added
2. istio-cni is ready
3. pods scheduled on this node by k8s

The sequence is guaranteed by taint controller. With this, `istio-cni` could catch each pods event and make sure all traffic could be intercepted ASAP without any exception 

Think further, without https://github.com/istio/istio/issues/40303, (2) and (3) would be reversed. Thus, before `istio-cni` ready, traffic will escape until `ztunnel` is ready.

So I think the main concern is whether we need do a immediate reconciliation while `istio-cni` startup rather than being triggered by `ztunnel` pod event?

Reply:
> I think #40303 is to guarantee `istio-cni` is ready before pods being scheduled on that node?

Yes - waiting for `istio-cni` readiness as mentioned in #40303 isn't enough anymore if we want to prevent pods from being scheduled before redirection AND proxying are ready - since gating on `istio-cni` readiness doesn't guarantee the latter at all (or even really the former - we might need to check both in all cases).


## Scenario 1: `istio-cni` "holds/waits" (via node taint or whatever else) until `ztunnel` is also ready on that node:

- To guarantee no traffic escapes -> we need to wait for `istio-cni` readiness (which implicitly also waits for ztunnel readiness).
- To guarantee pods are not scheduled until traffic _can be successfully proxied_ -> same as above.

 ## Scenario 2: `istio-cni` does NOT "hold/wait" (via node taint or whatever else) until `ztunnel` is also ready on that node:

- To guarantee no traffic escapes -> we need to wait for `istio-cni` readiness on that node.
- To guarantee pods are not scheduled until traffic _can be successfully proxied_ -> we need to wait for `ztunnel` readiness on that node, separately via some other mechanism.

Reply:
There're 2 points we're talking and thinking about: 1. Traffic escaping. 2. Traffic READY
Node taint is more meaningful for preventing traffic escaping, but traffic READY is based on AND of both cni and ztunnel.

So theoretically if we guarantee Pods scheduled after cni ready, we could make sure no escaping, but for some scenario like cni upgrade crash, it's still possible we have a gap interval (with plugin removed and no watcher), so reconciliation at cni startup looks more reasonable.

It looks OOS for the original change.

The original change is to reduce the possibility of traffic escaping. I think it's not harmful and let's move forward.

Reply:
Removing `s.delZtunnelEbpfOnNode()` from `nodeCleanup` ~is~ sort of harmful, though - right? That's my main problem with this PR.

- it means prog redirection is left in place in a cluster on uninstall (not current behavior) leaving clusters in a broken state on uninstall.

- it means prog redirection is left in place on uninstall (not current behavior) meaning if you reinstall to the same cluster and switch to `iptables` (a common testing scenario) network redirection will be badly broken.

Reply:
Modified the previous change a bit regarding current stage: 
Cleanup all ebpf probes while cni terminating (this should introduce traffic escape for upgrade case, we may track it in #43642)

Reply:
That sounds good to me, thank you!

Reply:
@howardjohn , any other comment or idea for it?

Reply:
I don't really like this tbh, it will result in downtime during onboarding. One of the key value propositions is that enabling ztunnel doesn't break anything.

Reply:
Hmm.. if "availability" is prior to security, I'll revert this piece of change for enforcing to incept traffic for this "on-ramp/off-ramp" state

Reply:
Security is important too, thats why I suggested it should only be on initial onboarding. Once its ready once its permanent. We do not have to compromise on either.